### ISSUE TITLE
OpenCL 2.0 Shared Virtual Memory Performance

### ISSUE THREAD

**Issue by bdenhollander** [openmm/openmm#3947]
I stumbled on this Intel article on [Shared Virtual Memory](https://www.intel.com/content/www/us/en/developer/articles/technical/opencl-20-shared-virtual-memory-overview.html) SVM and implemented it using the fine-grained buffer variant with additional assistance from this ARM [example](https://developer.arm.com/documentation/101574/0400/OpenCL-2-0/Shared-virtual-memory). Fine-grained buffer with atomics is not currently supported on my RX 6600. The results are quite quite good, surpassing the flush tests in https://github.com/openmm/openmm/issues/3937#issuecomment-1414723720.

|            | Original | SVM     | Change |
|------------|----------|---------|--------|
| gbsa       | 414.542  | 761.506 | 83.70% |
| rf         | 439.182  | 573.641 | 30.62% |
| pme        | 297.885  | 397.689 | 33.50% |
| apoa1rf    | 111.136  | 139.21  | 25.26% |
| apoa1pme   | 79.1162  | 95.0357 | 20.12% |
| apoa1ljpme | 65.5816  | 72.9115 | 11.18% |
| amoebagk   | 1.39965  | 1.39589 | -0.27% |
| amoebapme  | 4.56433  | 4.96588 | 8.80%  |

This approach avoids `enqueueReadBuffer`, which creates a DMA request on Windows. Now it waits on `findBlocksWithInteractions` to complete and set a value in `pinnedCountMemory[0]`.

![image](https://user-images.githubusercontent.com/44237618/217102518-42d95389-8fb8-4faf-adb3-b62505c70b6d.png)

Intel and AMD quickly released drivers supporting OpenCL 2.0 in 2014 but Nvidia never went beyond [evaluation](https://en.wikipedia.org/wiki/OpenCL#Timeline_of_vendor_implementations). Nvidia did announce support for [OpenCL 3.0](https://developer.nvidia.com/blog/nvidia-is-now-opencl-3-0-conformant/). Apple stopped supporting OpenCL after 1.2. As a result, this optimization is most relevant to AMD and Intel GPUs on Linux and Windows. The code is available in https://github.com/bdenhollander/openmm/commit/444f9835e12fd07658f9431421a0b9defa4b5811 and is setup so that a `USE_OPENCL_120` define could be used to avoid OpenCL 2.0 on Macs.

**Comment by philipturner** [comment]
Shared Memory would be possible with a Metal hipSYCL backend. I could also experiment with memory optimizations in a native Metal backend. But with OpenCL this advanced feature would be inaccessible on my personal computer.

**Comment by peastman** [comment]
We might be able to make this work.  The complication is that it uses a newer feature in host code.  Device code gets compiled at runtime, so it's easy to enable or disable features based on whether they're available.  To do it in host code, we need to decide at compile time.  If we compile in newer API calls, it will be impossible to run against any OpenCL 1.2 implementation.  Even if we don't use the features, the library won't even load.

NVIDIA announced OpenCL 3.0 support in April 2021.  I'm not sure what driver version it was added in.  If someone's driver is older than that, they won't even be able to load the OpenCL platform.  Usually we try to provide backward compatibility longer than that.  On clusters, people may not be able to upgrade the driver.  On the other hand, if someone has an NVIDIA GPU, they're probably using CUDA instead, so this is less of an issue.

Apple doesn't support it, but that's already a different package.  Mac packages would be compiled for OpenCL 1.2, and Linux and Windows ones would be compiled for 2.x.

**Comment by peastman** [comment]
It looks like OpenCL 3.0 was added in the 465.x driver branch, which was a prerelease.  The earliest supported driver to support it is 470.x, which was included with CUDA 11.4.

**Comment by philipturner** [comment]
OpenCL 3.0 doesn't equate to shared virtual memory. v3.0 made a lot of v2.0 features optional, including SVM. There's no upgrade going from v2.0 to v3.0, only permission for a downgrade. Hence why AMD is still on OpenCL 2.1. Nvidia's v3.0 could be the same v1.2 but _technically_ v3.0.

**Comment by peastman** [comment]
Right, but you can still load a library that links to the unsupported features.  And there's an API to check at runtime whether they're supported.  In @bdenhollander's implementation:

```c++
#ifndef USE_OPENCL_120
                    cl_device_svm_capabilities caps;
                    cl_int status = clGetDeviceInfo(devices[i](), CL_DEVICE_SVM_CAPABILITIES, sizeof(cl_device_svm_capabilities), &caps, NULL);
                    supportsSvm = status == CL_SUCCESS && (caps & CL_DEVICE_SVM_FINE_GRAIN_BUFFER);
#endif
```

**Comment by philipturner** [comment]
So if we were to perform the flushing optimization on M1, and instead perform the SVM optimization on AMD, that would be fine. Both platforms benefit with a comparable percent increase. Seems fair since both work by the same principle.

Benchmark | M1 Speedup | RDNA Speedup |
-- | -- | --- |
gbsa | +67% | +84% |
rf | +38% | +31% |
pme | +24% | +34% |
apoa1rf | +12% | +25% |
apoa1pme | +8% | +20% |
apoa1ljpme | +7% | +11% |
amoebagk | +0% | +0% |
amoebapme | +1% | +9% |
amber20-dhfr | +26% | n/a |
amber20-cellulose | +3% | n/a |
amber20-stmv | +1% | n/a |

M1 benefits less because its threadgroup memory is slower; do the math and this causes a slower %increase when considering only the flushing optimization. 5+2 -> 5 is a less drastic change than 4+2 -> 4.

**Comment by bdenhollander** [comment]
I found a bug in my implementation via `TestOpenCLGBSAOBCForce`. I updated some parts to use `pinnedCountMemory[0]` but  there are still a few areas using `getInteractionCount()` to  pull it from `OpenCLArray interactionCount`.

**Comment by bdenhollander** [comment]
The test is passing now on my [branch](https://github.com/openmm/openmm/compare/master...bdenhollander:openmm:OpenCL-2.0-SVM). On master, `enqueueReadBuffer` takes a snapshot of `interactionCount[0]` and then retrieves it after forces are calculated. My first version was using SVM allocated `pinnedCountMemory[0]` instead of cl::Buffer `interactionCount[0]`. Other kernels had changed the value stored in `pinnedCountMemory[0]` from 0 after `findBlocksWithInteractions` ran and the test failed. The latest version adds a specific parameter named `pinnedCount` for grabbing a snapshot while keeping the existing device-side `interactionCount` buffer unchanged.

Updated benchmarks:
|            | Original | Fixed SVM | Change  |
|------------|----------|---------|---------|
| gbsa       | 414.542  | 590.431 | 42.43%  |
| rf         | 439.182  | 594.011 | 35.25%  |
| pme        | 297.885  | 409.448 | 37.45%  |
| apoa1rf    | 111.136  | 143.773 | 29.37%  |
| apoa1pme   | 79.1162  | 95.905  | 21.22%  |
| apoa1ljpme | 65.5816  | 73.5134 | 12.09%  |
| amoebagk   | 1.39965  | 1.39963 | 0.00%   |
| amoebapme<br>#3885  | 4.56433  | 4.07881 | -10.64% |

**Comment by peastman** [comment]
Here's what I get on RX 5700 XT with the latest version of your branch.

|Test|Old|New|Ratio
|---|---|---|---|
gbsa|392.243|441.891|1.13
rf|564.75|567.805|1.01
pme|316.873|319.674|1.01
apoa1rf|146.6|154.087|1.05
apoa1pme|102.971|105.334|1.02
apoa1ljpme|79.9142|81.753|1.02
amoebagk|1.05267|1.04948|1.00
amoebapme|5.19928|6.33287|1.22
amber20-dhfr|347.547|357.6|1.03
amber20-cellulose|20.6122|20.6856|1.00
amber20-stmv|5.08028|5.07567|1.00

The good: it's never slower than the old version (except for trivial differences that are probably just noise).

The bad: the only benchmarks to get significant speedups are gbsa and amoebapme.  All others are only a few percent or less.

The confusing: my results are totally different from yours!

I'll try adding in one or more flushes as well to see what that does.

**Comment by philipturner** [comment]
With this optimization, more powerful GPUs get a larger speedup. It was already running faster on @bdenhollander's GPU, therefore the speedup would be larger.

Think of the timeline. Say you have 10s wait and 2s of actual simulation. This optimization jumps from 12s to 2s - 6x faster. Then imagine you have a better GPU, 10s wait and 1s of actual computation. The jump from 11s to 1s is 11x faster.

**Comment by bdenhollander** [comment]
We may also be seeing lower GPU utilization tests preferring boost clock vs. high GPU utilization tests favouring shaders and TDP.

![desmos-graph](https://user-images.githubusercontent.com/44237618/217397814-244770df-95ec-41ea-884e-8d6a5d4e9052.png)

**Comment by peastman** [comment]
Maybe, but that doesn't seem likely to explain it.  You're seeing huge speedups of 30-40% on benchmarks where I see almost no change at all.  Then there's amoebapme where I get a large speedup and you get a large slowdown.  I need to see if I can cobble together some way of profiling to get a look at what's happening.

**Comment by ex-rzr** [comment]
I'm not sure the current approach with 
```cpp
#ifdef USE_SVM
    pinnedCount[0] = interactionCount[0];
#endif
```
is completely correct. There is no guarantee that `pinnedCount` will contain the latest value of `interactionCount`.

**Comment by bdenhollander** [comment]
> Maybe, but that doesn't seem likely to explain it. You're seeing huge speedups of 30-40% on benchmarks where I see almost no change at all. Then there's amoebapme where I get a large speedup and you get a large slowdown.

This branch includes 1 flush (Perm 5), which contributes part of the gains. I run benchmarks with Windows power profile set to High Performance so the CPU is running at all core turbo. This keeps the CPU ready to respond without needing to boost from idle and helps a lot on gbsa and rf. What's more difficult to explain is why running 2 GROMACS CPU threads of Folding@Home actually improves OpenMM performance. The max clock speed decreases by about 200 MHz in the scenario. High Performance may keep the CPU clock speed up but maybe it doesn't prevent clock gating of areas of the CPU so running Folding@Home keeps everything alive.

![Background CPU Load](https://user-images.githubusercontent.com/44237618/217713384-43a0b798-de26-4a29-bc30-5722958b0944.png)

**Comment by peastman** [comment]
It looks like you've found a new way to optimize OpenMM: run GROMACS at the same time! :)

I'm afraid we've now passed into the realm of absurdity.  I really have no idea how to explain that!  The obvious answer would be that the extra load prevents the CPU from switching to a lower power state, but since you've already configured it to prevent that, I have no idea.  I'm going to tentatively conclude that there's something really messed up either in the driver or in Windows' thread scheduling and power management.

**Comment by bdenhollander** [comment]
@peastman  I benchmarked several scenarios to try to figure out why SVM appears to help more on Windows than Linux.

![SVM Scenarios](https://user-images.githubusercontent.com/44237618/221723999-1e0e81c2-495b-4430-b26f-dfb586d3902a.png)

Adding SVM alone does not help much as shown above. In some benchmarks it helps while in others it hurts.

Original + perm 5 flush is the same optimization as https://github.com/openmm/openmm/pull/3973. This [did not help much](https://github.com/openmm/openmm/issues/3937#issuecomment-1416377734) on your 5700 XT on Linux.

Combining perm 5 flush and SVM boosts gbsa, rf, and pme on my RX 6600 on Windows. No boost from perm 5 on Linux would end up similar to just adding SVM.

Original + Max Flush is perm 5 flush and removing the Windows specific flushes from CommonKernels.cpp. 

Combining SVM + perm 5 flush and removing the extra flushes provides the biggest boost on Windows. Since perm 5 didn't help much on Linux and removing Windows specific flushes would have no effect on Linux, this scenario would also be similar to just adding SVM.

Therefore, SVM makes other gap minimizing optimizations on Windows more effective. Some gaps were already not present in Linux and we haven't found the most effective place to shrink the gaps that do exist on Linux.

**Comment by bdenhollander** [comment]
SVM is not working consistently well on Linux compared to Window. Perhaps the driver pushing the value from the GPU back to main memory has a lower priority then an explicit enqueued read request so it ends up being slower on less complex systems. Given the requirement of OpenCL 2.0 and superiority of HIP for AMD, I am closing this issue.

![image](https://user-images.githubusercontent.com/44237618/230747369-4304f017-1633-4a31-81f7-117875681221.png)
